### ISSUE TITLE
Elastic search - how to test

### ISSUE THREAD

**Issue by ushah001** [thorsten/phpMyFAQ#1890]
**Describe the bug**
I have created index in elastic search configuration page. When I search with less than 4 characters, it returns data. For letters more than 4 character, I am getting error. 

**To Reproduce**

1. Go to elastic search configuration on Admin side and click Create Index, 
2. Go to Search page on front-end
3. Search for string less than and greater than 4 characters
4. Enable debug mode and see error

**Expected behavior**
Search should happen without errors. 

**Screenshots**
![image](https://user-images.githubusercontent.com/67292745/107661560-635e2180-6c57-11eb-9b42-c94268a77f53.png)

**phpMyFAQ (please complete the following information):**

- phpMyFAQ version - 3.0.5
- PHP version - 7
- Database - mySQL
- Elasticsearch yes/no - Yes 7.9.2

**Additional context**
I am not sure how to know if my search is working through elastic search or full-text.

**Comment by thorsten** [comment]
What happens if you disable the debug mode?

**Comment by thorsten** [comment]
What language do you use?

**Comment by ushah001** [comment]
en only.

**Comment by ushah001** [comment]
nothing happens. only white screen appears.

**Comment by thorsten** [comment]
Looks like nothing's returned for more than 4 letters.

**Comment by thorsten** [comment]
hm, what happens if you use the "find as you type" feature? Does it work?

**Comment by thorsten** [comment]
And please upgrade to phpMyFAQ 3.0.7

**Comment by ushah001** [comment]
If I disable elastic search from Edit Configuration --> Search then "find as you type" feature works and search works as well. 
![image](https://user-images.githubusercontent.com/67292745/107677720-9361f080-6c68-11eb-81b6-4bbfd6a09357.png)
![image](https://user-images.githubusercontent.com/67292745/107677765-9c52c200-6c68-11eb-91c5-275f86b5e680.png)

And in my organization, we needed few custom checkboxes and that's why we are not upgrading to 3.0.7 as we have to code again after the upgrade.

**Comment by thorsten** [comment]
Some errors you got are fixed with 3.0.7. What kind of custom checkboxes do you need? Are they interesting for all users?

**Comment by ushah001** [comment]
custom code changes like - 
1) new checkbox added for Pee review as my organization needed peer review and then publishing , same way for legal review
2) change in admin side report - added which user did published the FAQs, if FAQ is still active or no etc. 

I will try to update on my local and see. 

I have one question. Although I am very much new to elastic search. If I create index using phpmyfaq platform only, do I need to do any other code change/config change to index my data so that It starts searching using elastic search ? 

What I am unable to understand is, even after index is created - how data in my knowledge base and elastic search is connected. 

Sorry for my very basic questions.

**Comment by ushah001** [comment]
After upgrading to 3.0.7, it is still showing blank page when I search: 

![image](https://user-images.githubusercontent.com/67292745/107808759-13a35700-6d38-11eb-8f1b-043682eee3f1.png)

![image](https://user-images.githubusercontent.com/67292745/107808792-1f8f1900-6d38-11eb-93d6-f1471902ff48.png)

![image](https://user-images.githubusercontent.com/67292745/107808806-261d9080-6d38-11eb-861b-4d069f905532.png)

**Comment by thorsten** [comment]
You have to run a "full import" on the Elasticsearch page. Otherwise nothing is in the search index.

**Comment by thorsten** [comment]
@ushah001 The two feature mentioned are feature that I could integrate into 3.1 for all users, they sound useful.

**Comment by ushah001** [comment]
@thorsten , thanks for including these features to phpmyfaq. 
Further, when I did full import, I got below screen as 'undefined' and no errors. 

![image](https://user-images.githubusercontent.com/67292745/108086890-dcd58580-7044-11eb-961d-d0def33408b9.png)

and search is still showing me blank screen. what should appear when I do full import ?

**Comment by thorsten** [comment]
I think you missed one fix for the autocomplete stuff, did you change all JS files during the last updates? The wrong behaviour for the auto-completion was fixed in one of the 3.0.x releases and it was in a JS file. I'll check the PHP error when you turned on the debug mode.

**Comment by ushah001** [comment]
@thorsten , it looks like any word which is not part of the FAQ is giving me the error. not just less than 4 character. 
I debugged the code and found that issue is with Elasticsearch.php --> search function. 

when matching FAQs are not found, it should have returned No FAQs found. 

Could you please confirm this. 

![image](https://user-images.githubusercontent.com/67292745/108254322-dae4f280-7128-11eb-8176-9ec6be11c89e.png)

**Comment by thorsten** [comment]
Well, on my test systems it returns "no FAQs found", I cannot reproduce the error you got. Is it possible to get your database dump?

**Comment by ushah001** [comment]
sure anyway it is my localhost test db only. 
where should i attach it ? it is .sql file.

**Comment by thorsten** [comment]
You can send it to thorsten AT phpmyfaq DOT de

**Comment by ushah001** [comment]
hey @thorsten , finally I figured out the problem. 

when there is no matching result, the output from Elasticsearch.php --> search function --> $result = $this->client->search($searchParams); [line no 101] is as below: 

![image](https://user-images.githubusercontent.com/67292745/108262329-cad21080-7132-11eb-8bfb-fe5f9646480f.png)

when there is a match, the output is as below: 
![image](https://user-images.githubusercontent.com/67292745/108262420-e89f7580-7132-11eb-92e8-98e1377ca4e3.png)

Now, the line of code:  $result['hits']['total'] is never empty when there is no match. 

I changed the line to $result['hits']['hits'] and it gives me correct output. 

![image](https://user-images.githubusercontent.com/67292745/108262487-fd7c0900-7132-11eb-81fd-8ccb27388df4.png)

![image](https://user-images.githubusercontent.com/67292745/108262621-26040300-7133-11eb-8796-5bff60dd5bf6.png)

![image](https://user-images.githubusercontent.com/67292745/108262859-7f6c3200-7133-11eb-9b9e-e5566937658e.png)

all this happens, when elastic search support is enabled. 

![image](https://user-images.githubusercontent.com/67292745/108262987-a62a6880-7133-11eb-8986-bd0901158f93.png)

**Comment by thorsten** [comment]
Ha, and I know, why I couldn't reproduce it, as I fixed it by accident:

https://github.com/thorsten/phpMyFAQ/blob/main/phpmyfaq/src/phpMyFAQ/Search/Elasticsearch.php#L106

**Comment by ushah001** [comment]
Great. one last question before I close this, it is regarding full import. 

Just to clarify, do I need to do full-import every time as questions are updated or added every day on my side. So do I have to do full-import only once for the very first time after index is created in elastic search or regularly I have to do full-import ?

**Comment by thorsten** [comment]
The full import is only needed for the first time or e.g. if you want to clean the Elasticsearch index. If you add or update FAQs and you enabled the Elasticsearch feature, new or updated FAQs will be also added to Elasticsearch.
